pkg/icingaredis/telemetry/heartbeat.go

```diff
@@ -0,0 +1,102 @@
+package telemetry
+
+import (
+	"context"
+	"github.com/go-redis/redis/v8"
+	"github.com/icinga/icingadb/internal"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/icingaredis"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/periodic"
+	"github.com/icinga/icingadb/pkg/types"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/zap"
+	"strconv"
+	"time"
+)
+
+// StartHeartbeat periodically writes heartbeats to Redis for being monitored by Icinga 2.
+func StartHeartbeat(
+	ctx context.Context, client *icingaredis.Client, logger *logging.Logger,
+	instanceId types.Binary, envId *com.Atomic[types.Binary], db *sqlx.DB, dbErr *com.Atomic[error],
+) {
+	type ha struct {
+		isSet bool
+
+		active, inactive, weAreActive uint8
+	}
+
+	haInfo := &com.Atomic[ha]{}
+
+	periodic.Start(ctx, time.Second, func(_ periodic.Tick) {
+		values := map[string]string{"version": internal.Version}
+		if err, _ := dbErr.Load(); err != nil {
+			values["db-err"] = err.Error()
+		}
+
+		if ha, _ := haInfo.Load(); ha.isSet {
+			values["ha:active"] = strconv.FormatUint(uint64(ha.active), 10)
+			values["ha:inactive"] = strconv.FormatUint(uint64(ha.inactive), 10)
+			values["ha:we-are-active"] = strconv.FormatUint(uint64(ha.weAreActive), 10)
+		}
+
+		cmd := client.XAdd(ctx, &redis.XAddArgs{
+			Stream: "icingadb:telemetry:heartbeat",
+			MaxLen: 1,
+			Values: values,
+		})
+		if err := cmd.Err(); err != nil && !utils.IsContextCanceled(err) {
+			logger.Warnw("Can't update own heartbeat", zap.Error(icingaredis.WrapCmdErr(cmd)))
+		}
+	})
+
+	query := db.Rebind(
+		"SELECT COUNT(*) Amount, (CASE WHEN id = ? THEN 1 ELSE 0 END) IsMe, " +
+			"(CASE WHEN responsible = ? AND heartbeat > ? THEN 1 ELSE 0 END) IsResponsible " +
+			"FROM icingadb_instance WHERE environment_id = ? GROUP BY IsMe, IsResponsible",
+	)
+	periodic.Start(ctx, time.Second, func(_ periodic.Tick) {
+		envId, _ := envId.Load()
+		if envId == nil {
+			return
+		}
+
+		var rows []struct {
+			Amount        uint8 `db:"Amount"`
+			IsMe          uint8 `db:"IsMe"`
+			IsResponsible uint8 `db:"IsResponsible"`
+		}
+
+		// Give up after 3s, not 5m (default) to deliver (missing) HA info ASAP even if DB is down.
+		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+
+		err := db.SelectContext(
+			ctx, &rows, query, instanceId, "y", types.UnixMilli(time.Now().Add(-time.Minute)), envId,
+		)
```

Comment:
I don't like how this duplicates detection of active instances from `ha.go`, including a hardcoded copy of the timeout value. I'd rather prefer the normal heartbeat transaction to additionally return that extra information instead.

The query to fetch active instances could be extended to return more verbose information if desired. But in my opinion, the HA transaction already has enough information. Just having the two healthy states "this instance is active" and "another instance is active but this instance can take over at any time" should suffice.

Reply:
Too bad _the HA transaction_ runs only on heartbeat which in turn – if at all – fires 1x/3s.